cumulusci/oauth/client.py

```diff
@@ -0,0 +1,293 @@
+        {
+            "access_token":"<acces_token>",
+            "token_type":"bearer",
+            "expires_in":3600,
+            "refresh_token":"<refresh_token>",
+            "scope":"create"
+        }
+        """
+        print("auth_code_flow()")
+        auth_uri_with_params = self._get_auth_uri()
+        # Open a browser and direct the user to login
+        webbrowser.open(auth_uri_with_params, new=1)
+        # Open up an http deamon to listen for the
+        # callback from the auth server
+        self.httpd = self._create_httpd(use_https=use_https)
+        logger.info(
+            f"Spawning HTTP server at {self.client_info.callback_url}"
+            f" with timeout of {self.httpd.timeout} seconds.\n"
+            "If you are unable to log in to Salesforce you can"
+            " press <Ctrl+C> to kill the server and return to the command line."
+        )
+        # Implement the 300 second timeout
+        timeout_thread = HTTPDTimeout(self.httpd, self.httpd_timeout)
+        timeout_thread.start()
+        # use serve_forever because it is smarter about polling for Ctrl-C
+        # on Windows.
+        #
+        # There are two ways it can be shutdown.
+        # 1. Get a callback from Salesforce.
+        # 2. Timeout
+
+        try:
+            # for some reason this is required for Safari (checked Feb 2021)
+            # https://github.com/SFDO-Tooling/CumulusCI/pull/2373
+            old_timeout = socket.getdefaulttimeout()
+            socket.setdefaulttimeout(3)
+            self.httpd.serve_forever()
+        finally:
+            socket.setdefaulttimeout(old_timeout)
+
+        if use_https:
+            Path("key.pem").unlink()
+            Path("localhost.pem").unlink()
+
+        # timeout thread can stop polling and just finish
+        timeout_thread.quit()
+        self.validate_response(self.response)
+        return safe_json_from_response(self.response)
+
+    def _get_auth_uri(self):
+        url = self.client_info.auth_uri
+        url += "?response_type=code"
+        url += f"&client_id={self.client_info.client_id}"
+        url += f"&redirect_uri={self.client_info.callback_url}"
+        if self.client_info.scope:
+            url += f"&scope={quote(self.client_info.scope)}"
+        if self.client_info.prompt:
+            url += f"&prompt={quote('login')}"
+        return url
+
+    def _create_httpd(self, use_https=False):
+        """Create an http deamon process to listen
+        for the callback from the auth server"""
+        print(f"_create_httpd(use_https={use_https})")
+        url_parts = urlparse(self.client_info.callback_url)
+        server_address = (url_parts.hostname, url_parts.port)
+        OAuthCallbackHandler.parent = self
+
+        httpd = HTTPServer(server_address, OAuthCallbackHandler)
+        if use_https:
+            if not Path("localhost.pem").is_file() or not Path("key.pem").is_file():
+                create_pem_files()
```

Comment:
We should ideally create these in a temporary directory and clean them up when we're done.

Reply:
They are removed them [here](https://github.com/SFDO-Tooling/CumulusCI/pull/2602/files#diff-a5b3fa9198c38279d7a0e7a84e5cdfbe8ddf309fecbdf18c5f14c47750240623R144).

Reply:
That should probably go inside the finally block so they are removed even if there's an error